odl/operator/default_ops.py

```diff
@@ -1316,47 +1346,384 @@ def __init__(self, space):
         >>> op([1, -2])
         rn(2).element([ 1.,  2.])
 
-        The operator also works on other `FnBase` spaces such as
-        `DiscreteLp` spaces:
+        The operator also works on other `TensorSpace`'s such as
+        `DiscreteLp`:
 
         >>> r2 = odl.uniform_discr(0, 1, 2, dtype=complex)
         >>> op = odl.ComplexModulus(r2)
         >>> op([3 + 4j, 2])
         uniform_discr(0.0, 1.0, 2).element([ 5.,  2.])
         """
         real_space = space.real_space
-        linear = (space == real_space)
-        super(ComplexModulus, self).__init__(space, real_space, linear=linear)
+        super(ComplexModulus, self).__init__(space, real_space, linear=False)
 
     def _call(self, x):
         """Return ``self(x)``."""
         return (x.real ** 2 + x.imag ** 2).ufuncs.sqrt()
 
-    @property
-    def inverse(self):
-        """Return the (pseudo-)inverse.
+    def derivative(self, x):
+        r"""Return the derivative operator in the "C = R^2" sense.
+
+        The returned operator (``self``) is the derivative of the
+        operator variant where the complex domain is reinterpreted as
+        a product of two real spaces.
+
+        Parameters
+        ----------
+        x : `domain` element
+            Point in which to take the derivative.
 
         Examples
         --------
-        The (pseudo-)inverse in the real case is the identity:
+        >>> c2 = odl.cn(2)
+        >>> op = odl.ComplexModulus(c2)
+        >>> op([3 + 4j, 2])
+        rn(2).element([ 5.,  2.])
+        >>> deriv = op.derivative([3 + 4j, 2])
+        >>> deriv.domain
+        cn(2)
+        >>> deriv.range
+        rn(2)
+        >>> deriv([2 + 1j, 4j])  # [(3*2 + 4*1) / 5, (2*0 + 0*4) / 2]
+        rn(2).element([ 2.,  0.])
+
+        Notes
+        -----
+        The derivative of the complex modulus
+
+        .. math::
+            &M: X(\mathbb{C}) \to X(\mathbb{R}), \\
+            &M(x) = \sqrt{\Re(x)^2 + \Im(x)^2},
+
+        with :math:`X(\mathbb{F}) = \mathbb{F}^n` or
+        :math:`L^2(\Omega, \mathbb{F})`, is given as
+
+        .. math::
+            &M'(x): X(\mathbb{C}) \to X(\mathbb{R}), \\
+            &M'(x)(y) = \frac{\Re(x)\,\Re(y) + \Im(x)\,\Im(y)}{M(x)}.
+
+        It is linear when identifying :math:`\mathbb{C}` with
+        :math:`\mathbb{R}^2`, but not complex-linear.
+        """
+        op = self
+        x = self.domain.element(x)
+
```

Comment:
Do we want to special case x=0?

Reply:
That would be per component, and stuff like `x[np.isnan(x)] = 0` can add quite some overhead. I can do some benchmarking, but in principle I'm fine with NaN here. That also gives opportunity for debugging by filtering Numpy divide-by-zero errors.